text/3727-unsafe-set-discriminant.md

```diff
@@ -0,0 +1,299 @@
+
+Without this capability, downstream library authors are required to make a choice between:
+
+* Suboptimal stack usage and performance, in the case of by-value initializations
+* Restriction of functionality to a subset of enum reprs, or bespoke proxy types that emulate enums to allow for in-place construction
+
+This is necessary to be part of the language and/or standard library as enum representations (at least for the default `repr(Rust)`) are unspecified, and particularly with concepts like niche-packed enums, it is not possible to soundly handle this in the general case of user defined enum types.
+
+I am not aware of any other proposed designs for this functionality.
+
+# Prior art
+[prior-art]: #prior-art
+
+The `rkyv` crate defines [proxy enums] with explict `#[repr(N)]` (where N is an unsigned integer) types.
+
+[proxy enums]: https://rkyv.org/format.html
+
+Syntax for `discriminant_of!` is inspired by the in-progress [offset_of_enum feature][feature_offset_of_enum].
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## Should we use `&mut MaybeUninit<T>` for `set_discriminant`?
+
+There was some discussion when writing this RFC what the type of the first argument to `set_discriminant` should be:
+
+* `&mut MaybeUninit<T>`
+* `NonNull<T>`
+* `*mut T`
+
+From the perspective of "making expectations clear", I think `&mut MaybeUninit<T>` is the best choice, as it makes the requirements around alignment, exclusive access, etc. enumerated above explicit.
+
+However, it is possible that when constructing multiple nested types in-place, the user could be creating the pointers to the enum using `offset_of!` or `addr_of_mut!`, which would require the user to do some casting to turn their `*mut ()` (or any other `void*` analog) into a maybeuninit just to create an exclusive reference to it.
+
+We could use a "weaker" pointer type like `NonNull<T>` or `*mut T` instead, and cover the requirements in the "SAFETY" comment blocks, rather than "inheriting" these requirements from `&mut MaybeUninit<T>`.
+
+We could also have multiple versions of `set_discriminant`, like `set_discriminant_nonnull` or `set_discriminant_ptr` for these.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Should `discriminant_of!` support nested usage?
+
+Discussed above, we said this was not allowed:
+
+```rust
+discriminant_of!(Option<Option<u32>>, Some);        // Allowed
+discriminant_of!(Option<Option<u32>>, Some.0.Some); // Not Allowed (yet)
+```
+
+Should we allow this in the future?
+
+## Should we support construction of unaligned enums?
+
+If the user is writing to a packed format, they could potentially want the ability to set the discriminant in cases where the enum discriminant or fields are not well-aligned.
+
+This could require the creation of a `set_discriminant_unaligned` function, that relaxes the well-aligned safety requirements of the proposed `set_discriminant`.
```

Comment:
Personally, I would expect `mem::set_discriminant` to be paired with a `ptr::set_discriminant` that accepts a pointer instead of a reference, and so it would logically make sense that there also be a `ptr::set_discriminant_unaligned` as well.

Reply:
FWIW `set_discriminant_unaligned` is not something that currently exists in the compiler at all.

Note that `read/write_unaligned` are implemented entirely in the library, they don't need compiler support.

So supporting unaligned discriminant writes would be a much bigger change from the implementation side. I'd keep this as a future possibility for now. (And if we get  unaligned discriminant writes we surely should also get  unaligned discriminant reads.)

Reply:
Ah, that's fair. I would imagine that it would be possible to do with an unaligned read followed by an unaligned write, but it would require a bit of extra support from the compiler to ensure that you only read/write the part that contains the discriminant, rather than the entire type (which could be quite large).

Reply:
Noting that I don't have opinions about where the methods end up, I would defer to the desires of the lib/lang teams.

Reply:
I'll also add a note to "future potential" about unaligned discriminant reads as well.

Reply:
I mostly point out the option to have multiple methods because it feels like a very reasonable answer to "should this method accept a pointer or a reference" is "why not both?" since there's definitely precedent to do this. You're right that it's not super important right now, but it lends itself to the fact that the actual signature isn't nearly as important to the RFC as the actual functionality, since we can offer either or both signatures in the final version.